backend/cmd/headlamp.go

```diff
@@ -896,9 +900,112 @@ func (c *HeadlampConfig) OIDCTokenRefreshMiddleware(next http.Handler) http.Hand
 	})
 }
 
+// OIDCImpersonateMiddleware will validate and exchange the authorization JWT from the header with
+// the kubernetes service account JWT and instead impersonate the user using the `Impersonate-User`
+// header.
+// The headlamp service account must have the corresponding roles, see
+// https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation
+// If the token cannot be validated, the request will be forwarded to the next handler.
+func (c *HeadlampConfig) OIDCImpersonateMiddleware(next http.Handler) http.Handler {
+	oidcProvider, err := oidc.NewProvider(context.Background(), c.oidcIdpIssuerURL)
+	if err != nil {
+		logger.Log(logger.LevelError, map[string]string{"idpIssuerURL": c.oidcIdpIssuerURL},
+			err, "failed to get oidc provider")
+		return next
+	}
+
+	c.oidcProvider = oidcProvider
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// skip if not cluster request
+		if !strings.HasPrefix(r.URL.String(), "/clusters/") {
+			next.ServeHTTP(w, r)
+			return
+		}
+
+		// parse cluster and token
+		cluster, token := parseClusterAndToken(r)
+		if cluster == "" || token == "" {
+			next.ServeHTTP(w, r)
+			return
+		}
+
+		idToken, err := c.getValidToken(r.Context(), token)
+		if err != nil {
+			next.ServeHTTP(w, r)
+			return
+		}
+
+		var claims map[string]interface{}
+		if err := idToken.Claims(&claims); err != nil {
+			logger.Log(logger.LevelWarn, nil, err, "token claims could not be extracted")
+			next.ServeHTTP(w, r)
+			return
+		}
+		user, ok := claims[c.oidcUserClaim].(string)
+		if !ok || user == "" {
+			logger.Log(logger.LevelWarn, nil, err, "no user found in token")
+			next.ServeHTTP(w, r)
+			return
+		}
+		var groups []string
+		if c.oidcGroupsClaim != "" {
+			switch v := claims[c.oidcGroupsClaim].(type) {
+			case string:
+				groups = []string{v}
+			case []string:
+				groups = v
+			}
+		}
+
+		context, err := c.kubeConfigStore.GetContext(cluster)
+		if err != nil {
+			logger.Log(logger.LevelError, map[string]string{"cluster": cluster}, err,
+				"no serviceaccount token found")
+			next.ServeHTTP(w, r)
+			return
+		}
+
+		// User was successfully authenticated, execute request as this user
+		r.Header.Set("Authorization", "Bearer "+context.BearerToken)
+		// Since we are doing requests with the serviceaccount, make sure there are no
+		// other Impersonate-* headers added
+		for key, _ := range r.Header {
+			if strings.HasPrefix(strings.ToLower(key), "impersonate-") {
+				r.Header.Set(key, "")
+			}
+		}
+		r.Header.Set("Impersonate-User", user)
+		for _, group := range groups {
+			r.Header.Set("Impersonate-Group", group)
```

Comment:
I believe this should use Add instead of Set.

```suggestion
			r.Header.Add("Impersonate-Group", group)
```